### ISSUE TITLE
Generate wheel tags that reflect the bitness of the native executables

### ISSUE THREAD

**Issue by miguelsousa** [adobe-type-tools/afdko#352]
[From Cosimo](https://github.com/adobe-type-tools/afdko/pull/340#issuecomment-385931028),
> ideally, the afdko build system would build 32 bit binaries under a 32-bit python, and 64-bit binaries for 64-bit python (and "fat" 32+64 binaries on macOS). Currently it builds a mix of 32 or 64 bit binaries, and the generated wheel tags do not reflect the bitness of the native executables that are bundled with it.
The setup.py script could get the size of pointer from the current python interpreter, e.g. `import struct; bits = struct.calcsize('P') * 8)`, and then pass the right compiler flags to the makefiles.

**Comment by miguelsousa** [comment]
@anthrotype We just noticed that currently all the wheel files also include the source code files (*.c, *.h). Is this in any way related to what you pointed out above, or is it a separate thing?

**Comment by anthrotype** [comment]
Separate. It’s the include_package_data=True perhaps. Is that a problem?

**Comment by miguelsousa** [comment]
I don't know. We're just wondering if the wheel is unnecessarily bloated by containing those files. What purpose do they serve?

**Comment by readroberts** [comment]
@msousa There is no need for the wheel to contain the source files. You can leave out the "Programs" directory and everything under it.

**Comment by cjchapman** [comment]
@anthrotype & @miguelsousa, I stepped through the code that actually creates the bdist wheel file and it just recursively includes everything under the root (which in our case is the `afdko/afdko` folder). As far as I can see in the code (i.e. `make_wheelfile_inner()` in `wheel/archive.py`) There are no mechanisms for excluding any files, so I think we’d have to move the C/H source tree out to be under `afdko/` (i.e. parallel with `afdko/afdko/`) if we want to exclude it.

**Comment by anthrotype** [comment]
maybe also try adding `exclude_package_data`
http://setuptools.readthedocs.io/en/latest/setuptools.html#new-and-changed-setup-keywords

**Comment by anthrotype** [comment]
more info here: http://setuptools.readthedocs.io/en/latest/setuptools.html#including-data-files

**Comment by cjchapman** [comment]
@anthrotype I already tried setting `include_package_data=False`, it didn't seem to have any effect, that's why I stepped through the code, this is where I found that it's just walking the whole tree:
https://github.com/pypa/wheel/blob/3dc261abc98a5e43bc7fcf5783d080aaf8f9f0cf/wheel/archive.py#L60

**Comment by cjchapman** [comment]
@anthrotype I'll try `exclude_package_data` but I'm not optimistic, given the code I linked above.

**Comment by cjchapman** [comment]
@anthrotype I tried this just now:
```python
          include_package_data=False,
          exclude_package_data={'': ['*.c', '*.h']},
```
...but it did not appear to have any effect.

**Comment by anthrotype** [comment]
I didn't have the time to try, but I meant using exclude_package_data in addition to include_package_data=True, not setting include_package_data=False.
I don't think it's the wheel module that is responsible for determining what files are going to be installed. It's setuptools somewhere. Wheel just packages things up after setuptools has built and configured everything. Anyway, I wouldn't recommend adding further setup.py hacks to work around that, you should be able to achieve what you want by passing the right options to setuptools.
Try that exclude_package_data, or the more fine grained `package_data`.

**Comment by cjchapman** [comment]
@anthrotype OK, I tried this just now:
```python
          include_package_data=True,
          exclude_package_data={'': ['*.c', '*.h']},
```
...also no luck.

I note that the `include_package_data` documentation says:
>If set to True, this tells setuptools to automatically include any data files it finds inside your package directories that are specified by your MANIFEST.in file. For more information, see the section below on Including Data Files.

...but we're not currently using a `MANIFEST.in` file.

I'll look into `package_data` now.

**Comment by cjchapman** [comment]
@anthrotype I just tried this:
```python
          # include_package_data=True,
          # exclude_package_data={'': ['*.c', '*.h']},
          package_data={
              '': ['*.py'],
          },
```
...which I hoped would mean "only include *.py files", but that didn't work.

**Comment by anthrotype** [comment]
> exclude_package_data={'': ['*.c', '*.h']}

like `package_data`, it should be a dictionary that maps from package names to lists of glob patterns.
The glob patters are not recursive, so in the setuptools docs example, if you want to include (or exclude) all *.dat files under the data subdirectory of the mypkg package, you say

> 'mypkg': ['data/*.dat']

a package is a directory with an  `__init__.py`. So, for example in your case, `afdko` is a packge name, but also `afdko.Tools` is a (sub) package because the `afdko/Tools` directory contains an `__init__.py` (not sure what's the purpose of that particular one though, as I don't see other python modules inside that folder..).

**Comment by anthrotype** [comment]
this worked for me, although I had to remove the `./build` directory to make it work

```python
          include_package_data=True,
          exclude_package_data={'afdko': ['.gitignore']},
```

if I add that, then when I bdist_wheel, and unzip the wheel the afdko/.gitignore file is no longer there.

You know, if you have to maintain this complicated list of globs just to exclude your *.c and *.h and whatever other type of non-python file that you may add in the future... this is like maintaining a MANIFEST.in but in the reverse.
Why don't you simply move away in some sibling folders all the data files that you don't want to install alongside the afdko package?
Just add an `src` folder, inside have a `afdko` package (this one is the real one, which one would import from), and also other folders containing the C programs source files, and other data files, etc.

**Comment by miguelsousa** [comment]
I've opened #383 to track the removal of the source files from the binary wheels.

**Comment by miguelsousa** [comment]
We're going to try using [cibuildwheel](https://github.com/joerick/cibuildwheel) to address this issue.

**Comment by miguelsousa** [comment]
Also look into [wheelhouse-uploader](https://github.com/ogrisel/wheelhouse-uploader) for the deployment to PyPI.

**Comment by miguelsousa** [comment]
Fixed via #571 and #581
